src/base/bittorrent/private/sequentialdownloadsguard.cpp

```diff
@@ -0,0 +1,127 @@
+#include "sequentialdownloadsguard.h"
+
+#include <QDebug>
+
+#include "base/bittorrent/session.h"
+#include "base/bittorrent/torrenthandle.h"
+#include "base/logger.h"
+#include "base/utils/fs.h"
+#include "base/utils/misc.h"
+
+using namespace BitTorrent;
+
+/**
+   @brief SequentialDownloadsGuard explicit Constructor.
+ */
+SequentialDownloadsGuard::SequentialDownloadsGuard(Session *const session, QTimer *const sequentialGuardTimer) : QObject(session)
+    , m_session(session)
+    , m_sequentialGuardTimer(sequentialGuardTimer)
+{
+}
+
+/**
+   @brief Recheck Sequential download for all active torrents.
+ */
+void SequentialDownloadsGuard::recheckSequentialDownloads()
+{
+    // Increase QTimer interval after first shot
+    increaseTimerInterval();
+
+    foreach (TorrentHandle *const torrentHandle, m_session->torrents())
+        if (torrentHandle->isActive() && torrentHandle->isDownloading())
+            updateSequentialDownload(torrentHandle);
+}
+
+/**
+   @brief Enable / Disable Sequential download by torrents availability, torrent has to be previewable
+   to enable sequential download.
+ */
+void SequentialDownloadsGuard::updateSequentialDownload(TorrentHandle *const torrentHandle)
+{
+    // TODO: tmp to remove before merge PR, so other contributors can check the behavior silver
+    qDebug() << torrentHandle->name() << ":"
+             << "sd" << (torrentHandle->isSequentialDownload() ? "[ON]" : "[OFF]")
+             << "flp" << (torrentHandle->hasFirstLastPiecePriority() ? "[ON]" : "[OFF]")
+             << "sdo" << (torrentHandle->isSequentialDownloadOverride() ? "[ON]" : "[OFF]");
+
+    if (isSequentialDownloadEnabled(torrentHandle) &&
+        (torrentHandle->distributedCopies() > AVAILABILITY_LIMIT) &&
```

Comment:
Not strictly correct condition. The correct one has to account only pieces we are interested in, i.e. in a multi-file torrent if you download a single file, only its availability has to be taken in account.

Reply:
You are right, it doesn't take into account this and will not problem to check this per file, if libtorrent offers this info, but do we really need such precision? Here is important that torrent has enough seeders, so users can download pieces randomly ( or I don't know if there is any algorithm ).
But I like the idea to do it like you wrote ;)

Reply:
@silverqx:  

>if libtorrent offers this info 

*Libtorrent* does not offer per-file `availability` values, but *qBittorrent* now does, thanks to @evsh and his pieces-bar work.  

An implementation example can be seen here:  [<https://github.com/qbittorrent/qBittorrent/pull/4994>].

Reply:
yeah, looks nice, thx for link will be really helpful